stl/inc/semaphore

```diff
@@ -0,0 +1,304 @@
+                const auto _Remaining_timeout = __std_atomic_wait_get_remaining_timeout(_Deadline);
+                if (_Remaining_timeout == 0) {
+                    return false;
+                }
+                _Wait(_Remaining_timeout);
+                _Current = _Counter.load(memory_order_relaxed);
+            }
+            _STL_VERIFY(_Current > 0 && _Current <= _Least_max_value,
+                "Invariant: counter >= 0, and counter <= max() "
+                "possibly caused by preconditions violation (N4861 [thread.sema.cnt]/8)");
+
+            if (_Counter.compare_exchange_weak(_Current, _Current - 1, memory_order_seq_cst)) {
+                return true;
+            }
+        }
+    }
+
+    template <class _Clock, class _Duration>
+    _NODISCARD bool try_acquire_until(const chrono::time_point<_Clock, _Duration>& _Abs_time) {
+        ptrdiff_t _Current = _Counter.load(memory_order_relaxed);
+        for (;;) {
+            while (_Current == 0) {
+                const unsigned long _Remaining_timeout = _Semaphore_remaining_timeout(_Abs_time);
+                if (_Remaining_timeout == 0) {
+                    return false;
+                }
+                _Wait(_Remaining_timeout);
+                _Current = _Counter.load(memory_order_relaxed);
+            }
+            _STL_VERIFY(_Current > 0 && _Current <= _Least_max_value,
+                "Invariant: counter >= 0, and counter <= max() "
+                "possibly caused by preconditions violation (N4861 [thread.sema.cnt]/8)");
+
+            if (_Counter.compare_exchange_weak(_Current, _Current - 1, memory_order_seq_cst)) {
+                return true;
+            }
+        }
+    }
+
+private:
+    atomic<ptrdiff_t> _Counter;
+    atomic<ptrdiff_t> _Waiting;
+};
+
+template <>
+class counting_semaphore<1> {
+public:
+    _NODISCARD static constexpr ptrdiff_t(max)() noexcept {
+        return 1;
+    }
+
+    constexpr explicit counting_semaphore(const ptrdiff_t _Desired) noexcept /* strengthened */
+        : _Counter(static_cast<int8_t>(_Desired)) {
+        _STL_VERIFY((_Desired & ~1) == 0, "Precondition: desired >= 0, and desired <= max() "
+                                          "(N4861 [thread.sema.cnt]/5)");
+    }
+
+    counting_semaphore(const counting_semaphore&) = delete;
+    counting_semaphore& operator=(const counting_semaphore&) = delete;
+
+    void release(const ptrdiff_t _Update = 1) noexcept /* strengthened */ {
+        if (_Update == 0) {
+            return;
+        }
+        _STL_VERIFY(_Update == 1, "Precondition: update >= 0, "
+                                  "and update <= max() - counter (N4861 [thread.sema.cnt]/8)");
+        // TRANSITION, GH-1133: should be memory_order_release
+        _Counter.store(1, memory_order_seq_cst);
+        _Counter.notify_one();
+    }
+
+    void acquire() noexcept /* strengthened */ {
+        for (;;) {
+            // TRANSITION, GH-1133: should be memory_order_acquire

```

Comment:
No this one certainly needs to be acq_rel since we are publishing the 0 to other threads.

Reply:
acquire is enough, though it is publishing 0, doesn't need to strongly happen before, only release before acquire, not acquire before another acquire

Reply:
Hmmm OK.